tables/TaQL/ExprNodeSetOpt.h

```diff
@@ -0,0 +1,358 @@
+  // to determine if a value is contained in one of the intervals.
+  // <br>This templated class (as well as its derived classes) are instatiated
+  // for Double and String.
+  // </synopsis>
+
+  template <typename T>
+  class TableExprNodeSetOptContSet: public TableExprNodeSetOptContSetBase
+  {
+  public:
+    TableExprNodeSetOptContSet (const TableExprNodeSet& orig,
+                                const std::vector<T>& starts,
+                                const std::vector<T>& ends,
+                                const std::vector<Bool>& leftC,
+                                const std::vector<Bool>& rightC);
+    TableExprNodeSetOptContSet (const TableExprNodeSet& orig,
+                                const std::vector<T>& starts,
+                                const std::vector<T>& ends);
+    // Show the node.
+    virtual void show (ostream& os, uInt indent) const override;
+    // Tell which interval contains a value. -1 = no match.
+    virtual Int64 find (T value) const override;
+    // Transform a set into an optimized one by ordering the intervals
+    // and optionally combining adjacent intervals.
+    // If not possible, an empty TENShPtr is returned.
+    // It fill <src>rowNrs</src> with the row numbers of the intervals created.
+    static TENShPtr transform (const TableExprNodeSet& set,
+                               Bool combine=True);
+    // Create the appropriate optimized OptContSet object.
+    // Note that leftC and rightC do not need to have the same length as start/end.
+    // If it is known that all intervals have the same leftC/rightC,
+    // a single value suffices.
+    static TENShPtr createOptSet (const TableExprNodeSet& set,
+                                  const std::vector<T>& start,
+                                  const std::vector<T>& end, 
+                                  const std::vector<Bool>& leftC,
+                                  const std::vector<Bool>& rightC);
+    // Get the size (nr of intervals),
+    size_t size() const
+      { return itsStarts.size(); }
+  protected:
+    std::vector<T> itsStarts;
+    std::vector<T> itsEnds;
+    std::vector<Bool> itsLeftC;
+    std::vector<Bool> itsRightC;
+  };
+
+
+  // <summary>
+  // An optimized representation of a selection set with closed-closed intervals.
+  // </summary>
+
+  // <use visibility=local>
+
+  // <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="">
+  // </reviewed>
+
+  // <prerequisite>
+  //# Classes you should understand before using this one.
+  //   <li> TableExprNodeSetOptContSet
+  // </prerequisite>
+
+  // <synopsis>
+  // This class is a further optimized version of TableExprNodeSetOptContSet
+  // for continuous intervals all using a closed start and closed end.
+  // It reduces the number of comparisons required.
+  // </synopsis>
+
+  template <typename T>
+  class TableExprNodeSetOptContSetCC: public TableExprNodeSetOptContSet<T>
```

Comment:
Why inherit from `TableExprNodeSetOptContSet` instead of `TableExprNodeSetOptContSetBase`.
It saves writing the `size()` function, but it stores two unused Bool vectors.

However that generic part could be move to the Base and make then `itsStarts` and `itsEnds` protected members.

Reply:
itsStarts and itsEnds cannot be moved to ContSetBase because they are templated, while ContSetBase is not.
The leftC and rightC vectors are empty for those classes, thus not a real problem I think.